### PR TITLE
 Implement a DuckDuckGo Lite engine

### LINKED ISSUE #349: Implement DuckDuckGo-Lite engine (incl. paging)
Cross request: https://github.com/searx/searx/issues/2982

I can take the time to implement that, but if someone else wants to do it then feel free to do it.

### PR DESCRIPTION
**Pull request by return42**
## What does this PR do?

Implements a DuckDuckGo Lite engine

## Why is this change important?

The existing [DuckDuckGo engine](https://github.com/searxng/searxng/blob/master/searx/engines/duckduckgo.py) does not support paging. DuckDuckgo-Lite is much faster, [less verbose](https://news.ycombinator.com/item?id=22077861) and does have a paging option.

## How to test this PR locally?

    !ddg paris

Test date ranges and pages

## Related issues

Closes https://github.com/searxng/searxng/issues/349

### CODE REVIEW

File: searx/engines/duckduckgo_lite.py
Comment (dalf) [inline searx/engines/duckduckgo_lite.py]: searching for `Paris` in `tr-TR` language, the engine crash.

Possible fix:
```python
        a_tag = eval_xpath_getindex(tr_title, './/td//a[@class="result-link"]', 0, None)
        if not a_tag:
            continue
```

File: searx/engines/duckduckgo_lite.py
Comment (dalf) [inline searx/engines/duckduckgo_lite.py]: Resiliant version with a default empty string:
```
td_content = eval_xpath_getindex(tr_content, './/td[@class="result-snippet"]', 0, '')
```

File: searx/engines/duckduckgo_lite.py
Comment (dalf) [inline searx/engines/duckduckgo_lite.py]: Perhaps another PR: the last URL is https://duckduckgo.com/util/u588.js 

To get this URL:
* download `https://duckduckgo.com/`
* search for `<script type="text/javascript" src="/util/u([0-9]+).js"></script>`

**Review by dalf — CHANGES_REQUESTED** [review]
* There are some debug code here and there
* When to use duckduckgo.py ? When to use duckduckgo_lite.py ?

**Comment by return42** [comment]
> There are some debug code here and there

That is true, I gave some debug code to play around with the fields while we are in a testing phase / will be dropped before we merge.

> When to use duckduckgo.py ? When to use duckduckgo_lite.py ?

Its up to the instance maintainer I thought .. otherwise we can also drop duckduckgo.py if you want.

**Comment by unixfox** [comment]
> > There are some debug code here and there
> 
> That is true, I gave some debug code to play around with the fields while we are in a testing phase / will be dropped before we merge.
> 
> > When to use duckduckgo.py ? When to use duckduckgo_lite.py ?
> 
> Its up to the instance maintainer I thought .. otherwise we can also drop duckduckgo.py if you want.

If duckduckgo_lite.py offers the same functionality as duckduckgo.py then I don't see the point of keeping duckduckgo.py.

File: searx/engines/duckduckgo_lite.py
Comment (return42) [reply searx/engines/duckduckgo_lite.py]: see 3298fbc

File: searx/engines/duckduckgo.py
Comment (return42) [inline searx/engines/duckduckgo.py:105]: Today I improved the paging support, I try to emulate what the input forms from https://lite.duckduckgo.com/lite/  do when you click on "next page" .. while looking along the implementation the "API" seems a bit strange, but this is what the input forms do ...

**Comment by return42** [comment]
> If duckduckgo_lite.py offers the same functionality as duckduckgo.py then I don't see the point of keeping duckduckgo.py.

OK, I have removed the old implementation and reordered the patches into one commit, I also fixed the issues @dalf reported.

@unixfox do you have time to test (approve) DuckDuckGo-Lite?

**Comment by return42** [comment]
FYI: I rebased commit (c628a07c) "[upd] make data.languages" on #366

**Comment by dalf** [comment]
Same as the normal version, there is a ping url.
Without this ping, the results are not the same.

```python
url_ping = "https://duckduckgo.com/t/sl_l"


# get response from search-request
def response(resp):
    ...
    # ping
    headers_ping = dict_subset(resp.request.headers, ['User-Agent', 'Accept-Encoding', 'Accept', 'Cookie'])
    get(url_ping, headers=headers_ping)
```

**Comment by unixfox** [comment]
> Same as the normal version, there is a ping url. Without this ping, the results are not the same.
> 
> ```python
> url_ping = "https://duckduckgo.com/t/sl_l"
> 
> 
> # get response from search-request
> def response(resp):
>     ...
>     # ping
>     headers_ping = dict_subset(resp.request.headers, ['User-Agent', 'Accept-Encoding', 'Accept', 'Cookie'])
>     get(url_ping, headers=headers_ping)
> ```

What do you mean by the ping?

**Comment by dalf** [comment]
The HTML result page of lite.duckduckgo.com contains a 1 pixel image: ``.

See https://github.com/searx/searx/pull/2560

Same behavior, without a call to this URL, the second request get results with a lower quality.

**Comment by return42** [comment]
@dalf / thanks a lot, I haven't seen the `` in the ddg-lite page.  I fixed it https://github.com/searxng/searxng/pull/360/commits/59db8657e48b9becb5b6a9536ea0a1d9b0d8cd95#diff-7c54e676433f29906f915977e88587f6dc0eed69412bfd55324fa89540c28ccdR124-R125

**Comment by unixfox** [comment]
> The HTML result page of lite.duckduckgo.com contains a 1 pixel image: ``.
> 
> See https://github.com/searx/searx/pull/2560
> 
> Same behavior, without a call to this URL, the second request get results with a lower quality.
>  

Also when not using api=d.js and o=json?

File: searx/engines/duckduckgo.py
Comment (unixfox) [inline searx/engines/duckduckgo.py]: it's https://duckduckgo.com/t/sl_l instead

**Review by unixfox — CHANGES_REQUESTED** [review]

**Comment by unixfox** [comment]
Apart from the url ping it's working great, well on Linux, on my Mac not sure why but I'm getting this error:
```
DEBUG   searx.search                  : actual_timeout=10.0 (default_timeout=10.0, ?timeout_limit=None, max_request_timeout=None)
DEBUG   searx.engines.duckduckgo      : param data: {'q': 'ok', 's': 30, 'dc': 31, 'o': 'json', 'api': 'd.js', 'kl': 'us-en', 'df': ''}
DEBUG   searx.engines.duckduckgo      : param cookies: {'kl': 'us-en'}
DEBUG   searx.network.duckduckgo      : HTTP Request: POST https://lite.duckduckgo.com/lite "HTTP/2 200 OK" (text/html; charset=UTF-8)
DEBUG   searx.network.duckduckgo      : HTTP Request: GET https://duckduckgo.com/t/sl_l "HTTP/2 200 OK" (image/gif)
encoding error : input conversion failed due to input error, bytes 0x21 0x00 0x00 0x00
DEBUG   searx.webapp                  : http%3A//localhost%3A8888/search uses locale `fr`
INFO    werkzeug                      : 127.0.0.1 - - [01/Oct/2021 19:34:11] "POST /search HTTP/1.1" 200 -
```

**Comment by return42** [comment]
>  it's https://duckduckgo.com/t/sl_l instead

I need glasses / fixed / thanks!

> well on Linux, on my Mac not sure why but I'm getting this error:

On any request? .. I tried several requests but I can't reproduce this encoding error .. even if I try different languages.

**Comment by unixfox** [comment]
> > well on Linux, on my Mac not sure why but I'm getting this error:
> 
> On any request? .. I tried several requests but I can't reproduce this encoding error .. even if I try different languages.

When clicking on the next page button with the keyword "ok" and the language: English: http://localhost:8888/search?q=ok&time_range=&language=en&category_general=on

**Comment by return42** [comment]
> When clicking on the next page button with the keyword "ok" and the language: English

Hm .. I don't see this issue ... strange ..

**Comment by unixfox** [comment]
> > When clicking on the next page button with the keyword "ok" and the language: English
> 
> Hm .. I don't see this issue ... strange ..

Don't worry it's probably an issue related to my workflow, something went wrong with homebrew. that's ok I don't even need to run searx on my work mac, it was just for testing your PR, I usually use my linux pc for that.

**Review by unixfox — APPROVED** [review]

**Comment by dalf** [comment]
same as @return42, I can't reproduce the "input conversion failed due to input error" message.
Note: this is message comes from lxml not from HTTP get on  `https://duckduckgo.com/t/sl_l` = the error comes from `doc = fromstring(resp.text)`

**Comment by return42** [comment]
> the error comes from `doc = fromstring(resp.text)`

@dalf, @unixfox thanks for testing and all your feedback.  I'm going to merge this PR even if we can't clarify this encoding issue right now.  If we see this issue in practice we can fix it afterwards I think.

**update:** released on https://darmarit.org/searx

**Pull request merged**
